### DISCUSSION TITLE
limit() & offset() can not be adpated to prestosql

### DISCUSSION THREAD

**Discussion by Neo-J** [sqlalchemy/sqlalchemy#7708] [Usage Questions] [answered]
Hi guys,

I am trying to use sqlalchemy orm with prestosql. Here is the problem, my statement is like 
`db.query(Model).filter(a=xxx).offset(10).limit(10).all()`
and the query generated is like 
`select ... from table where ... LIMIT %(param_1)s OFFSET %(param_2)s
`
However, in presto, the limit...offset... statement should be used reversely
`select * from table offset 10 limit 10;`

If using the generated one, it will give you below error

> sqlalchemy.exc.DatabaseError: (pyhive.exc.DatabaseError) {'message': "line 4:10: mismatched input 'OFFSET'. Expecting: <EOF>", 'errorCode': 1, 'errorName': 'SYNTAX_ERROR', 'errorType': 'USER_ERROR', 'errorLocation': {'lineNumber': 4, 'columnNumber': 10}, 'failureInfo': {'type': 'io.prestosql.sql.parser.ParsingException', 'message': "line 4:10: mismatched input 'OFFSET'. Expecting: <EOF>", 'cause': {'type': 'org.antlr.v4.runtime.InputMismatchException', 'suppressed': [], 'stack': ['io.prestosql.sql.parser.SqlParser$2.recoverInline(SqlParser.java:130)', 'org.antlr.v4.runtime.Parser.match(Parser.java:206)', 'io.prestosql.sql.parser.SqlBaseParser.singleStatement(SqlBaseParser.java:249)', 'io.prestosql.sql.parser.SqlParser.invokeParser(SqlParser.java:156)', 'io.prestosql.sql.parser.SqlParser.createStatement(SqlParser.java:95)', 'io.prestosql.execution.QueryPreparer.prepareQuery(QueryPreparer.java:55)', 'io.prestosql.dispatcher.DispatchManager.createQueryInternal(DispatchManager.java:179)', 'io.prestosql.dispatcher.DispatchManager.lambda$createQuery$0(DispatchManager.java:148)', 'io.prestosql.$gen.Presto_340____20220214_144905_2.run(Unknown Source)', 'java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)', 'java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)', 'java.base/java.lang.Thread.run(Thread.java:834)']}, 'suppressed': [], 'stack': ['io.prestosql.sql.parser.ErrorHandler.syntaxError(ErrorHandler.java:108)', 'org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)', 'org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)', 'org.antlr.v4.runtime.DefaultErrorStrategy.reportInputMismatch(DefaultErrorStrategy.java:327)', 'org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:139)', 'io.prestosql.sql.parser.SqlBaseParser.singleStatement(SqlBaseParser.java:254)', 'io.prestosql.sql.parser.SqlParser.invokeParser(SqlParser.java:156)', 'io.prestosql.sql.parser.SqlParser.createStatement(SqlParser.java:95)', 'io.prestosql.execution.QueryPreparer.prepareQuery(QueryPreparer.java:55)', 'io.prestosql.dispatcher.DispatchManager.createQueryInternal(DispatchManager.java:179)', 'io.prestosql.dispatcher.DispatchManager.lambda$createQuery$0(DispatchManager.java:148)', 'io.prestosql.$gen.Presto_340____20220214_144905_2.run(Unknown Source)', 'java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)', 'java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)', 'java.base/java.lang.Thread.run(Thread.java:834)'], 'errorLocation': {'lineNumber': 4, 'columnNumber': 10}}}

Is there a way to fix or by pass it without using the customized statement?

**Comment by CaselIT** [comment, marked as answer]
Hi,

Sqlalchemy does not support prestosql directly. I'm assuming you are using an extension to add support for it. You should report this in the external dialect repository
